File: medbrains/crates/medbrains-core/src/specialty/psychiatry.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "psych_admission_category", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum PsychAdmissionCategory {
    Independent,
    Supported,
    MinorSupported,
    Emergency,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "ect_laterality", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum EctLaterality {
    Bilateral,
    RightUnilateral,
    LeftUnilateral,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "restraint_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum RestraintType {
    Physical,
    Chemical,
    Seclusion,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct PsychPatient {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub patient_id: Uuid,
    pub admission_category: PsychAdmissionCategory,
    pub advance_directive_text: Option<String>,
    pub nominated_rep_name: Option<String>,
    pub nominated_rep_relation: Option<String>,
    pub nominated_rep_contact: Option<String>,
    pub substance_abuse_flag: bool,
    pub is_restricted: bool,
    pub treating_psychiatrist_id: Option<Uuid>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct PsychAssessment {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub psych_patient_id: Uuid,
    pub assessment_type: String,
    pub mental_status_exam: serde_json::Value,
    pub ham_d_score: Option<i32>,
    pub bprs_score: Option<i32>,
    pub gaf_score: Option<i32>,
    pub risk_assessment: serde_json::Value,
    pub assessed_by: Uuid,
    pub assessed_at: DateTime<Utc>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct PsychEctRegister {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub psych_patient_id: Uuid,
    pub session_number: i32,
    pub consent_obtained: bool,
    pub laterality: EctLaterality,
    pub stimulus_dose: Option<rust_decimal::Decimal>,
    pub seizure_duration_sec: Option<i32>,
    pub anesthetic: Option<String>,
    pub muscle_relaxant: Option<String>,
    pub performed_by: Uuid,
    pub anesthetist_id: Option<Uuid>,
    pub session_date: DateTime<Utc>,
    pub complications: Option<String>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct PsychSeclusionRestraint {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub psych_patient_id: Uuid,
    pub restraint_type: RestraintType,
    pub reason: String,
    pub start_time: DateTime<Utc>,
    pub review_due_at: DateTime<Utc>,
    pub reviewed_at: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub ordered_by: Uuid,
    pub released_by: Option<Uuid>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct PsychMhrbNotification {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub psych_patient_id: Uuid,
    pub notification_type: String,
    pub reference_number: Option<String>,
    pub status: String,
    pub sent_at: Option<DateTime<Utc>>,
    pub acknowledged_at: Option<DateTime<Utc>>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct PsychCounselingSession {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub psych_patient_id: Uuid,
    pub session_type: String,
    pub therapist_id: Uuid,
    pub modality: Option<String>,
    pub duration_minutes: Option<i32>,
    pub outcome_rating: Option<i32>,
    pub session_date: DateTime<Utc>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

File: medbrains/packages/types/src/psychiatry.ts
// Psychiatry types — split from index.ts, barrel-re-exported.

// ── Psychiatry ──

export type PsychAdmissionCategory = "independent" | "supported" | "minor_supported" | "emergency";

export type EctLaterality = "bilateral" | "right_unilateral" | "left_unilateral";

export type RestraintType = "physical" | "chemical" | "seclusion";

export interface PsychPatient {
  id: string;
  tenant_id: string;
  patient_id: string;
  admission_category: PsychAdmissionCategory;
  advance_directive_text: string | null;
  nominated_rep_name: string | null;
  nominated_rep_contact: string | null;
  nominated_rep_relation: string | null;
  substance_abuse_flag: boolean;
  is_restricted: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreatePsychPatientRequest {
  patient_id: string;
  admission_category: PsychAdmissionCategory;
  advance_directive_text?: string;
  nominated_rep_name?: string;
  nominated_rep_contact?: string;
  nominated_rep_relation?: string;
  substance_abuse_flag?: boolean;
}

export interface PsychAssessment {
  id: string;
  tenant_id: string;
  psych_patient_id: string;
  assessment_type: string;
  mental_status_exam: Record<string, unknown> | null;
  ham_d_score: number | null;
  bprs_score: number | null;
  risk_assessment: Record<string, unknown> | null;
  created_at: string;
}

export interface CreatePsychAssessmentRequest {
  assessment_type: string;
  mental_status_exam?: Record<string, unknown>;
  ham_d_score?: number;
  bprs_score?: number;
  risk_assessment?: Record<string, unknown>;
}

export interface PsychEctSession {
  id: string;
  tenant_id: string;
  psych_patient_id: string;
  session_number: number;
  consent_obtained: boolean;
  laterality: EctLaterality;
  stimulus_dose: string | null;
  seizure_duration: string | null;
  anesthetic: string | null;
  performed_by: string;
  created_at: string;
}

export interface CreatePsychEctRequest {
  session_number: number;
  consent_obtained: boolean;
  laterality: EctLaterality;
  stimulus_dose?: string;
  seizure_duration?: string;
  anesthetic?: string;
  performed_by: string;
}

export interface PsychRestraint {
  id: string;
  tenant_id: string;
  psych_patient_id: string;
  restraint_type: RestraintType;
  start_time: string;
  review_due_at: string;
  reviewed_at: string | null;
  released_at: string | null;
  released_by: string | null;
  created_at: string;
}

export interface CreatePsychRestraintRequest {
  restraint_type: RestraintType;
  start_time: string;
}

export interface PsychMhrbNotification {
  id: string;
  tenant_id: string;
  psych_patient_id: string;
  notification_type: string;
  reference_number: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface CreatePsychMhrbRequest {
  notification_type: string;
  reference_number?: string;
}

export interface PsychCounselingSession {
  id: string;
  tenant_id: string;
  psych_patient_id: string;
  session_type: string;
  therapist_id: string;
  modality: string | null;
  outcome_rating: number | null;
  created_at: string;
}

export interface CreatePsychCounselingRequest {
  session_type: string;
  therapist_id: string;
  modality?: string;
  outcome_rating?: number;
}
